File: src/requests/events.ts
import { filter, map, share } from 'rxjs/operators'
import { webSocket } from "rxjs/webSocket"
import type { FolderViewItem } from '../components/FolderView'

export type CmdMsg = {
    cmd: string
}
export type CmdToggleMsg = {
    cmd: string,
    checked: boolean
}
export type StatusMsg = {
    folderId: string,
    requestId: number,
    text?: string
}
export type ExtendedInfo = {
    folderId: string,
    requestId: number,
    items: FolderViewItem[]
}

export type CopyProgress = {
    move: boolean
    name: string,
    totalCount: number,
    currentCount: number,
    totalMaxBytes: number,
    totalBytes: number,
    previousTotalBytes: number,
    currentMaxBytes: number,
    currentBytes: number,
    duration: number
}

type WebSocketMsg = {
    method: "cmd" | "cmdtoggle" | "status" | "extendedinfo" | "copyprogress",
    cmdMsg?: CmdMsg,
    cmdToggleMsg?: CmdToggleMsg,
    statusMsg?: StatusMsg
    extendedInfo?: ExtendedInfo,
    copyProgress?: CopyProgress,
    progressRunning?: boolean
}

const socket = webSocket<WebSocketMsg>('ws://localhost:20000/events').pipe(share())

export const cmdEvents = socket
                    .pipe(filter(n => n.method == "cmd"))
                    .pipe(map(n => n.cmdMsg)!)
export const cmdToggleEvents = socket
                    .pipe(filter(n => n.method == "cmdtoggle"))
                    .pipe(map(n => n.cmdToggleMsg!))
export const statusEvents = socket
                    .pipe(filter(n => n.method == "status"))
                    .pipe(map(n => n.statusMsg!))
export const exifDataEvents = socket
                    .pipe(filter(n => n.method == "extendedinfo"))
                    .pipe(map(n => n.extendedInfo!))
export const copyProgressEvents = socket
                    .pipe(filter(n => n.method == "copyprogress"))
                    .pipe(map(n => n.copyProgress!))

socket.subscribe({
    error: err => console.error('Subscription error:', err),
    complete: () => console.log('WebSocket connection closed'),
})

File: Commander/Events.cs
using Commander.Controllers;
using Commander.ProgressAction;
using WebServerLight;

namespace Commander;

static class Events
{
    public static void WebSocket(IWebSocket webSocket)
        => Events.webSocket = webSocket;

    public static async void SendMenuCommand(string id)
    {
        if (webSocket != null)
            await webSocket.SendJson(new WebSocketMsg("cmd", new(id)));
    }

    public static async void SendMenuCheck(string id, bool check)
    {
        if (webSocket != null)
            await webSocket.SendJson(new WebSocketMsg("cmdtoggle", null, new(id, check)));
    }

    public static async void SendStatusBarInfo(string id, int requestId, string? text)
    {
        if (webSocket != null)
            await webSocket.SendJson(new WebSocketMsg("status", StatusMsg: new(id, requestId, text)));
    }

    public static async void SendExtendedInfo(string id, int requestId, DirectoryItem[] items)
    {
        try
        {
            if (webSocket != null)
                await webSocket.SendJson(new WebSocketMsg("extendedinfo", ExtendedInfo: new(id, requestId, items)));
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error sending extended info: {ex}");
        }
    }

    public static async void SendCopyProgress(CopyProgress copyProgress)
    {
        if (webSocket != null)
            await webSocket.SendJson(new WebSocketMsg("copyprogress", CopyProgress: copyProgress));
    }

    public static async void SendRaw(string msg)
    {
        if (webSocket != null)
            await webSocket.SendString(msg);
    }

    static IWebSocket? webSocket;
}

record WebSocketMsg(
    string Method,
    CmdMsg? CmdMsg = null,
    CmdToggleMsg? CmdToggleMsg = null,
    StatusMsg? StatusMsg = null,
    ExtendedInfo? ExtendedInfo = null,
    CopyProgress? CopyProgress = null,
    bool? ProgressRunning = null
);

record CmdMsg(string Cmd);
record CmdToggleMsg(string Cmd, bool Checked);
record StatusMsg(
    string FolderId,
    int RequestId,
    string? Text);
record ExtendedInfo(
    string FolderId,
    int RequestId,
    DirectoryItem[] Items);

